tutorials/02-Basic Operations.ipynb

```diff
@@ -0,0 +1,1822 @@
+   "metadata": {},
+   "source": [
+    "## Functional Programming"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "464a5750-64f5-4083-b534-73f58183ebe5",
+   "metadata": {},
+   "source": [
+    "Because of the immutability of sympy expressions,\n",
+    "it is encouraged to do functional programming with sympy.\n",
+    "\n",
+    "You should treat every data as read-only, and manipulate data only in functions without any side effect.\n",
+    "\n",
+    "Because sympy expressions are immutable, sympy functions \n",
+    "can't introduce any side effects on the on the expressions,\n",
+    "which makes it very easy to understand and use sympy functions. \n",
+    "\n",
+    "(In fact, I encourage you to do functional programming \n",
+    "and think 'symbolically' in business to become a better programmer in general)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ecb022a3-9498-4a0b-98e0-4b765e7b19fc",
+   "metadata": {},
+   "source": [
+    "### Exercise"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d2b0e95d-a4bf-48f3-909d-230a3ce85176",
+   "metadata": {},
+   "source": [
+    "Chebyshev polynomials are defined by the recurrence relation\n",
+    "\n",
+    "$$T_k(x) = 2x T_{k-1}(x) - T_{k-2}(x)$$\n",
+    "\n",
+    "With initial conditions:\n",
+    "$$T_0(x) = 1$$\n",
+    "$$T_1(x) = x$$\n",
+    "\n",
+    "Build your own recursive function that can compute the chebyshev polynomials:\n",
+    "\n",
+    "(Use `chebyshev(k, x)` signature to pass the parameters `k, x`)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 134,
+   "id": "be293453-d144-41ec-a4cd-21ab8ef3bd07",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/latex": [
+       "$\\displaystyle 1$"
+      ],
+      "text/plain": [
+       "1"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "text/latex": [
+       "$\\displaystyle x$"
+      ],
+      "text/plain": [
+       "x"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "text/latex": [
+       "$\\displaystyle 2 x^{2} - 1$"
+      ],
+      "text/plain": [
+       "2*x**2 - 1"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "text/latex": [
+       "$\\displaystyle 2 x \\left(2 x^{2} - 1\\right) - x$"
+      ],
+      "text/plain": [
+       "2*x*(2*x**2 - 1) - x"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "text/latex": [
+       "$\\displaystyle - 2 x^{2} + 2 x \\left(2 x \\left(2 x^{2} - 1\\right) - x\\right) + 1$"
+      ],
+      "text/plain": [
+       "-2*x**2 + 2*x*(2*x*(2*x**2 - 1) - x) + 1"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "def chebyshev(k, x):\n",
+    "    match k:\n",
```

Comment:
Might be best to avoid newer Python syntax in case anybody has an older Python version.

Reply:
Anyway, I'm a little unclear, isn't this the solution to the previous exercise? Shouldn't it be omitted, or at least hidden?